SmartTutor/ContentModel/LearningObjects/ChallengeModel/FulfillmentStrategy/MetricChecker/BasicMetricsChecker.cs

```diff
@@ -33,19 +40,89 @@ private List<CaDETMember> GetMethodsFromClasses(List<CaDETClass> caDETClasses)
         private bool ValidateClassMetricRules(List<CaDETClass> caDETClasses)
         {
             foreach (CaDETClass caDETClass in caDETClasses)
+            {
                 foreach (MetricRangeRule classMetricRule in ClassMetricRules)
-                    if (!classMetricRule.MetricMeetsRequirements(caDETClass.Metrics))
-                        return false;
+                {
+                    if (!classMetricRule.MetricMeetsRequirements(caDETClass.Metrics)) return false;
+                }
+            }
             return true;
         }
 
         private bool ValidateMethodMetricRules(List<CaDETMember> caDETMethods)
         {
             foreach (CaDETMember caDETMethod in caDETMethods)
+            {
                 foreach (MetricRangeRule methodMetricRule in MethodMetricRules)
-                    if (!methodMetricRule.MetricMeetsRequirements(caDETMethod.Metrics))
-                        return false;
+                {
+                    if (!methodMetricRule.MetricMeetsRequirements(caDETMethod.Metrics)) return false;
+                }
+            }
             return true;
         }
+
+        private List<ChallengeHint> GetChallengeHints()
```

Comment:
The code below seems to have a lot of unneeded logic. Make sure you go over my comment below in the MetricRangeRuleClass first.

Then consider redesigning the logic in this class to have the following structure:
public override ChallengeEvaluation CheckChallengeFulfillment(List<CaDETClass> solutionAttempt)
{
    List<ChallengeHint> hints = new List<>();
    hints.AddRange(ValidateClassMetricRules(solutionAttempt));
    hints.AddRange(ValidateMethodMetricRules(solutionAttempt));
    if(hints.IsEmpty()) return new ChallengeEvaluation(true, GetAllHints());
    return new ChallengeEvaluation(false, hints);
}

The ValidateXMetricRules functions will create an empty hint list, go over the list of rules and for each rule that does not pass, ask it if it has a Hint - if yes add it, if no ignore it.

Reply:
What happens when there is no hint but metric value is not appropriate? If there is no hint

> The code below seems to have a lot of unneeded logic. Make sure you go over my comment below in the MetricRangeRuleClass first.
> 
> Then consider redesigning the logic in this class to have the following structure:
> public override ChallengeEvaluation CheckChallengeFulfillment(List solutionAttempt)
> {
> List hints = new List<>();
> hints.AddRange(ValidateClassMetricRules(solutionAttempt));
> hints.AddRange(ValidateMethodMetricRules(solutionAttempt));
> if(hints.IsEmpty()) return new ChallengeEvaluation(true, GetAllHints());
What happens when there is no hint but metric value is not appropriate? Is tutor required to put hint for every metric rule or is there need for another validation which returns bool?